### PR TITLE
Add requestedServerName() to Network::connection

### PR DESCRIPTION
**Pull request by vadimeisenbergibm**
*Description*: Add requestedServerName() to Network::connection, to enable using the SNI value in the filters. The current use case  is sending the SNI value to the Istio mixer for monitoring and policy enforcement. Suppose we have *.ibm.com value in sni_domains, and a request is sent to www.ibm.com, the requirement is to extract the value of www.ibm.com by a listener and to send it to Mixer.
*Risk Level*: Low
*Testing*: Added SslSocketTest.GetRequestedServerName test
*Docs Changes*: None
*Release Notes*: None
Is related to issue https://github.com/envoyproxy/envoy/issues/3707, is a prerequisite for https://github.com/istio/istio/issues/6810

### CODE REVIEW

File: test/common/ssl/ssl_socket_test.cc [hunk 1]
```diff
@@ -142,6 +142,7 @@ const std::string testUtilV2(const envoy::api::v2::Listener& server_proto,
                              const std::string& expected_client_cert_uri,
                              const std::string& expected_alpn_protocol,
                              const std::string& expected_stats, unsigned expected_stats_value,
+                             const std::string& expected_requested_server_name,
```
Comment (PiotrSikora) [inline test/common/ssl/ssl_socket_test.cc]: Nit, but could you move this before `expected_alpn_protocol`?

Comment (vadimeisenbergibm) [reply test/common/ssl/ssl_socket_test.cc]: @PiotrSikora Sure, doing it now.

**Comment by ggreenway** [comment]
Waiting to merge until @vadimeisenbergibm pushes the commit mentioned in latest comment

**Comment by vadimeisenbergibm** [comment]
@PiotrSikora @ggreenway Done, took me some time, sorry.

**Pull request merged**